### ISSUE TITLE
Setting fuse_conn_info->max_write on OpenSUSE Tumbleweed leads to "reading device: Invalid argument" error during mount

### ISSUE THREAD

**Issue by kiril-kirov** [libfuse/libfuse#512]
Hello, 

We've used `max_write` for pretty long time on plenty of different Linux distributions and it works fine - if the kernel is before 4.20, it just doesn't do anything. If it's newer - it's works like charm.
Only on OpenSUSE Tumbleweed, setting `max_write` to anything **larger** than 1 MiB leads to error during mount, stating `reading device: Invalid argument`.

This can easily be reproduced using the `hello` example, just setting `conn->max_write = 2 * 1024 * 1024`. Enabling the debug output shows:

```
> ./hello -d  ~/mnt/
FUSE library version: 3.6.2
nullpath_ok: 0
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.31
flags=0x03fffffb
max_readahead=0x00080000
   INIT: 7.29
   flags=0x0040f039
   max_readahead=0x00080000
   max_write=0x00200000
   max_background=0
   congestion_threshold=0
   time_gran=1
   unique: 2, success, outsize: 80
unique: 4, opcode: ACCESS (34), nodeid: 1, insize: 48, pid: 3621
fuse: reading device: Invalid argument
   unique: 4, error: -38 (Function not implemented), outsize: 16
```

More info:
```
> uname -a
Linux localhost.localdomain 5.5.11-1-default #1 SMP Sat Mar 21 11:11:19 UTC 2020 (794b2e9) x86_64 x86_64 x86_64 GNU/Linux

> cat /etc/os-release 
NAME="openSUSE Tumbleweed"
# VERSION="20200326"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20200326"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20200326"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
LOGO="distributor-logo"
```

The error is not present even on OpenSUSE Leap.
The same happens with libfuse 3.9.1.

A wild guess would be - some kind of incompatibility between kernel module and libfuse? Or the kernel does not support `max_pages` (although it's still not supposed to throw such errors in that case, e.g. latest Leap is with 4.12 and it doesn't complain; not to mention that the Tumbleweed kernel is 5.5.11).

My guess would be that some **additional check should be done during the capability negotiation phase, but what?** Didn't see anything interesting there.
What could possibly cause this error and how could it be avoided?

Thanks!

Kiril

**Comment by kiril-kirov** [comment]
Interestingly, setting `max_read` to 8MiB does't produce this error.

**Comment by kiril-kirov** [comment]
**TL;DR;** Seems like the issue is with kernels 5.5+, as I stumbled upon this problem with Debian 11 Alpha 2 as well.
Setting `max_write` to 1MiB seems to work fine. Also seems like the writes are capped to 1MiB at max, no matter what larger value `max_write` has.

**So, the main unanswered question is - is there a way to check what's the max allowed value and use it instead of hardcoding 1MiB?**

---

Tested on:
* Mint 19 (5.0.0-32)
* Ubuntu 19.10 (5.3.0-26)
* Fedora 31 (5.3.7-301)

with a simple test application, which directly makes POSIX calls (with `sleep`s between calls and `fsync`s) for:
* writes with size: 64KiB, 128KiB, 512KiB, 1MiB, 2MiB, 5MiB, 8MiB and 16MiB
* reads with size: 64KiB, 128KiB, 512KiB, 1MiB, 2MiB, 5MiB, 8MiB and 16MiB

Results:

* On all distros fuse write callbacks receive buffers with the expected size as long as they are **no larger than 1MiB**. Once the writes become larger, the kernel module (or the OS?) seems to split the writes, so for an 8MiB write, the callback is called 8 times with 1MiB buffers.
* On all distros fuse reads callbacks are called with the expected size as long as the reads are **no larger than 128KiB**. Once the reads become larger, the kernel module (or the OS?) seems to split the reads, so for an 1MiB read, the callback is called 8 times with 128KiB buffers.

Making `max_write` and `max_read` (for symmetry) 1MiB each, mount was possible on:
* Debian 11 (Bullseye, Alpha 2, 5.5.0-1)
* OpenSuSE Tumbleweed (5.5.11-1)

Executing the same test app on these two platforms led to the following:
* same result for writes
* same result for reads on Debian
* reads can be up to 512KiB on OpenSuse - cool!

**Comment by trapexit** [comment]
No. The cap is just hard coded to 256 in newer kernels and 32 in older 
ones. You get the FUSE_MAX_PAGES flag in the INIT message and can set it 
in the response but the INIT messages doesn't include the max value.
